orchestration/flows/bl832/nersc.py

```diff
@@ -0,0 +1,207 @@
+from dotenv import load_dotenv
+import logging
+import os
+from pathlib import Path
+from prefect import flow
+from typing import Optional
+
+from orchestration.flows.bl832.config import Config832
+from orchestration.flows.bl832.job_controller import get_controller, TomographyHPCController
+from orchestration.nersc import NerscClient
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+load_dotenv()
+
+
+class NERSCTomographyHPCController(TomographyHPCController):
+    """
+    Implementation for a NERSC-based tomography HPC controller.
+
+    Submits reconstruction and multi-resolution jobs to NERSC via SFAPI.
+    """
+
+    def __init__(
+        self,
+        client: NerscClient = None,
+        Config832: Optional[Config832] = None
+    ) -> None:
+        self.client = client
+        if not Config832:
+            self.config = Config832()
+        else:
+            self.config = Config832
+
+    def create_nersc_client() -> NerscClient:
+        """Create and return an NERSC client instance"""
+
+        client_id_path = os.getenv("PATH_NERSC_CLIENT_ID")
+        sfapi_key_path = os.getenv("PATH_NERSC_PRI_KEY")
+
+        if not client_id_path or not sfapi_key_path:
+            logger.error("NERSC credentials paths are missing.")
+            raise ValueError("Missing NERSC credentials paths.")
+        if not os.path.isfile(client_id_path) or not os.path.isfile(sfapi_key_path):
+            logger.error("NERSC credential files are missing.")
+            raise FileNotFoundError("NERSC credential files are missing.")
+
+        try:
+            return NerscClient(client_id_path, sfapi_key_path)
+        except Exception as e:
+            logger.error(f"Failed to create NERSC client: {e}")
+            raise e
+
+    def reconstruct(
+        self,
+        file_path: str = "",
+    ) -> bool:
+        """
+        Use NERSC for tomography reconstruction
+        """
+        logger.info("Starting NERSC reconstruction process.")
+
+        # Can't use this long term in production. Need to find a better way to handle credentials.
+        # Want to run this as the alsdev user
+        username = os.getenv("NERSC_USERNAME")
+        password = os.getenv("NERSC_PASSWORD")
+
+        user = self.client.user()
+
+        home_path = f"/global/homes/{user.name[0]}/{user.name}"
+        scratch_path = f"/pscratch/sd/{user.name[0]}/{user.name}"
+        logger.info(home_path)
+        logger.info(scratch_path)
+
+        image_name = self.config.harbor_images832["recon_image"]
+        path = Path(file_path)
+        folder_name = path.parent.name
+        file_name = path.stem
+
+        # Can't use this long term in production. Need to find a better way to handle credentials.
+        # Want to run this as the alsdev user
+        username = os.getenv("NERSC_USERNAME")
+        password = os.getenv("NERSC_PASSWORD")
+
+        job_script = f"""#!/bin/bash
+            #SBATCH -q preempt
```

Comment:
Several queue things I have figured out:

- `debug` has no minimum time requirement, but using a different queue such as `preempt` requires allocating 2 hours minimum for a job (which I think is a bit overkill in our case, since recon should only take 5-10 min).
- Based on testing debug vs preempt, it seems like the the debug jobs get picked up much faster (perhaps due to smaller time constraint, as they are both medium priority).
- There is also the `interactive` queue, which has a high priority, but only 4 nodes x 2 jobs can be active simultaneously.

Reply:
Bjoern let us know that we (under the 'als' project) have access to the `realtime` queue at NERSC, which should reduce latency. I had an issue using that queue with the following error:
`Error during job submission or completion: sbatch: error: Batch job submission failed: Invalid qos specification`

I opened a ticket and we are still waiting to hear from the help desk about the correct job parameters to use the queue:
https://nersc.servicenowservices.com/sp?sys_id=03d747c187a216101c6cea470cbb35e9&view=sp&id=ticket&table=incident